source_middleware.go

```diff
@@ -598,6 +626,341 @@ func (s *sourceWithSchemaContext) LifecycleOnDeleted(ctx context.Context, config
 	return s.Source.LifecycleOnDeleted(schema.WithSchemaContextName(ctx, s.contextName), config)
 }
 
-func ptr[T any](t T) *T {
-	return &t
+// -- SourceWithBatch --------------------------------------------------
+
+const (
+	configSourceBatchSize  = "sdk.batch.size"
+	configSourceBatchDelay = "sdk.batch.delay"
+)
+
+// SourceWithBatchConfig is the configuration for the
+// SourceWithBatch middleware. Fields set to their zero value are
+// ignored and will be set to the default value.
+//
+// SourceWithBatchConfig can be used as a SourceMiddlewareOption.
+type SourceWithBatchConfig struct {
+	// BatchSize is the default value for the batch size.
+	BatchSize *int
+	// BatchDelay is the default value for the batch delay.
+	BatchDelay *time.Duration
+}
+
+// Apply sets the default configuration for the SourceWithBatch middleware.
+func (c SourceWithBatchConfig) Apply(m SourceMiddleware) {
+	if d, ok := m.(*SourceWithBatch); ok {
+		d.Config = c
+	}
+}
+
+func (c SourceWithBatchConfig) BatchSizeParameterName() string {
+	return configSourceBatchSize
+}
+
+func (c SourceWithBatchConfig) BatchDelayParameterName() string {
+	return configSourceBatchDelay
+}
+
+func (c SourceWithBatchConfig) parameters() config.Parameters {
+	return config.Parameters{
+		configSourceBatchSize: {
+			Default:     strconv.Itoa(*c.BatchSize),
+			Description: "Maximum size of batch before it gets read from the source.",
+			Type:        config.ParameterTypeInt,
+		},
+		configSourceBatchDelay: {
+			Default:     c.BatchDelay.String(),
+			Description: "Maximum delay before an incomplete batch is read from the source.",
+			Type:        config.ParameterTypeDuration,
+		},
+	}
+}
+
+// SourceWithBatch adds support for batching on the source. It adds
+// two parameters to the source config:
+//   - `sdk.batch.size` - Maximum size of batch before it gets written to the
+//     source.
+//   - `sdk.batch.delay` - Maximum delay before an incomplete batch is written
+//     to the source.
+//
+// To change the defaults of these parameters use the fields of this struct.
+type SourceWithBatch struct {
+	Config SourceWithBatchConfig
+}
+
+// Wrap a Source into the batching middleware.
+func (s *SourceWithBatch) Wrap(impl Source) Source {
+	if s.Config.BatchSize == nil {
+		s.Config.BatchSize = lang.Ptr(0)
+	}
+	if s.Config.BatchDelay == nil {
+		s.Config.BatchDelay = lang.Ptr(time.Duration(0))
+	}
+
+	return &sourceWithBatch{
+		Source:   impl,
+		defaults: s.Config,
+
+		readCh:      make(chan readResponse, *s.Config.BatchSize),
+		readBatchCh: make(chan readBatchResponse, 1),
+		stop:        make(chan struct{}),
+	}
+}
+
+type sourceWithBatch struct {
```

Comment:
IIUC, the purpose of this middleware is to make sure that Conduit receives batches of records, of a certain size and with a certain delay?

Reply:
Yes, basically the same as the destination batch middleware (it even uses the same config parameters), only that the batching is happening on the source instead of the destination. The pipeline architecture v2 then processes this batch end-to-end as one unit, as opposed to the v1 architecture, which processes records one by one.